\documentclass[10pt,a4paper]{article}

\usepackage{amssymb}
\usepackage[french,english]{babel}
\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{lineno}
\usepackage{cite}
\usepackage{float}
\usepackage{ccaption}
\usepackage{caption}
\usepackage{array}
\usepackage{lscape}
\usepackage[hmargin=2cm,vmargin=2cm]{geometry}
\usepackage{fancyvrb}
\usepackage{hyperref}

\title{\bf \Large A short manual for {\tt TESS3}:
a program to estimate spatial population structure\\
\large (command-line and R wrapper)
}

\author{
        Kevin Caye (kevin.caye@imag.fr)\\
        Olivier FranÃ§ois (olivier.francois@imag.fr)\\
}

\newcommand{\bp}{\mathbf{p}}
\newcommand{\LLL}{\mathcal{L}}

%% BEGIN DOC
\begin{document}


\maketitle
\begin{center}
{\it Please, print this reference manual only if it is necessary.}
\end{center}

\noindent
This short manual aims to help users to run {\tt TESS3} command-line engine and R wrapper function on Mac, Linux and Windows.

\section{Description} 

Inference of spatial population structure are commonly perform with computer program based on intensive stochastic simulation. These methods do not scale with the dimension of the data sets generated from next-generation sequencing technologies. The computer program {\tt TESS3} has functionalities similar to the spatial bayesian clustering program  {\tt TESS}~\cite{durand2009spatial}, but has run-times several order faster than those of {\tt TESS} 2.3. The method is based on geographically constrained non-negative matrix factorization, and provides similar results that with {\tt TESS} 2.3. In addition {\tt TESS3} computes an ancestral allele frequency differentiation statistic that can be used to perform selection scans.

\section{Installation} 

The source code of {\tt TESS3} can be download on github: 
 \url{https://github.com/cayek/TESS3/archive/master.zip}.
Then you just need to follow instructions on the github project page \url{https://github.com/cayek/TESS3}.

\section{Data format}

\subsection{input file}
The {\tt TESS3} input file consists of a genotype file in the {\bf geno} format and coordinate file in the {\bf coord} format. 
\begin{itemize}
\item {\bf geno} (example.geno)

The {\bf geno} format has one row for each SNP.
  Each row contains 1 character per individual:
  0 means zero copies of the reference allele.
  1 means one copy of the reference allele.
  2 means two copies of the reference allele.
  9 means missing data.

Below, an example of a geno file for $n=3$ individuals and $L=4$ loci.
\begin{center}
\footnotesize
\begin{Verbatim}[frame=single]
112
010
091
121
\end{Verbatim}
\end{center}


\item {\bf coord} (example.coord)

The {\bf coord} format has one row for each individual. Each row contains the \verb|longitude latitude| information of the individual.

Below, an example of a geno file for $n=3$ individuals and $L=4$ loci.
\begin{center}
\footnotesize
\begin{Verbatim}[frame=single]
2.515455200203690111e+01 5.439077948590419709e+01
-8.429345306090160861e+00 4.019713388759510053e+01
1.351129055178800087e+01 5.585331731335860184e+01
\end{Verbatim}
\end{center}

\end{itemize}

\noindent
Other formats for genotype data sets can be used thanks to the {\tt LEA} R package~\cite{frichot2015lea}. Indeed, {\tt LEA} enable to convert into {\bf geno} format the following usual formats : {\bf ped}, {\bf ancestrymap}, {\bf vcf} and {\bf lfmm}.

\subsection{output files}
\noindent
There are three main {\bf output files}.

\begin{itemize}
\item The file with the extension the {\bf .Q} contains individual admixture coefficients.
It contains a matrix with $n$ rows (the number of individuals) and $K$ columns (the 
number of ancestral populations).
\item The file with the extension {\bf .G} contains the ancestral genotypic frequencies.
It contains a matrix with $n_a\times L$ lines (the number of alleles times the number of SNPs) 
and $K$ columns (the number of ancestral populations). 
For a diploid SNP, the first line contains the ancestral frequencies for the number 
of allele equals to 0, the second line contains the ancestral frequencies for the 
number allele equals to 1, the third line contains the ancestral frequencies for 
the number of alleles equal to 2.
\item The file with the extension {\bf .Fst} contains ancestral allele frequency differentiation statistic. In this file each row is the statistic estimate for each loci. 
\end{itemize}

There also are less important {\bf output files}, but that can be useful to have information on the graph computed, the least-squared criterion and the cross-entropy criterion.

\begin{itemize}
\item The file with the extension the {\bf .W} contains the edge weight matrix of the nearest-neighbor graph computed from coordinates data.
\item The file with the extension the {\bf .sum} contains the value of least-squared criterion, the cross entropy criteria on all data and only on masked data if the user asked it.
\end{itemize}

\section{Run the program}
\subsection{Command-line}
The {\tt TESS3} program can be executed from a command line. The format is:
\begin{Verbatim}[frame=single]
./TESS3 -x genotype_file.geno -K number_of_ancestral_populations  -r coordinates_file
\end{Verbatim}

\noindent
Only these three options are mandatory. There is no ordering for the options in the command line. 
Here is a description of these options:
\begin{itemize}
\item \verb|-x genotype_file.geno| is the path to the genotype file (in .geno format).
\item \verb|-K number_of_ancestral_populations| is the number of ancestral populations. 
\item \verb|-r coordinates_file| is the path to the coordinate file (in .coord format). 
\end{itemize}

\noindent
Additional options are available:
\begin{itemize}
\item \verb|-a alpha| is the value of the normalized regularization parameter (by default: 0.001). This parameter control the spatial regularity of the ancestry estimates.
\item \verb|-W edge_weight_input| is the path to the file that contains a edge weight matrix of the spatial graph. The file has to contain each element of the matrix separates by a space row by row. The program use this graph in place of the one computed if no edge weight input file is given.
\item \verb|-q output_Q| is the path for the output file containing the ancestry coefficients. By default, the name of the output file is the same name as the input file with the extension .K.Q.
\item \verb|-g output_G| is the path to the output file containing the ancestral genotype frequencies. By default, the name of the output file is the same name as the input genotype file with the extension .K.G.
\item \verb|-f output_FST| is the path to the output file containing the ancestral allele frequency differentiation statistic. By default, the name of the output file is the same name as the input genotype file with the extension .K.Fst.
\item \verb|-y output_FST| is the path to the output file containing the value of least-squared criterion, the cross entropy criteria on all data and only on masked data. By default, the name of the output file is the same name as the input genotype file with the extension .K.sum.
\item \verb|-c perc| is the percentage of masked genotypes. If this option is set, the cross-entropy criterion is calculated (see\cite{frichot2014fast} for more details on the cross-entropy criterion). The default percentage is $5\%$.
\item \verb|-e tolerance| is the tolerance error in the {\tt TESS3} optimization algorithm (by default: 0.0000001). 
\item \verb|-i iteration_number| is the max number of iterations of the algorithm (default: 200). 
\item \verb|-I nb_SNPs| starts the algorithm with a run of {\tt TESS3} using a subset of nb\_SNPs random SNPs. This option can speed up {\tt TESS3} estimation for very large data sets.
\item \verb|-Q input_Q| is the path to an initial file for the $Q$ matrix containing individual admixture coefficients. If both \verb|-I| and \verb|-Q| are set, \verb|-Q| is chosen.
\item \verb|-s seed| is a seed to initialize the random number generator. 
\item \verb|-m ploidy|  1 if haploid, 2 if diploid (default: 2). 
\item \verb|-p p| is the number of CPUs to use when the algorithm is run on a multiprocessor system.
Be aware that the number of processes has to be lower or equal to the number 
of CPU units available on your computer (default: 1).

\end{itemize}


\noindent
If you need a summary of options, you can use the \verb|-h| option by typing the following command
\footnotesize
\begin{Verbatim}[frame=single]
./TESS3 -h
\end{Verbatim}
\noindent
\normalsize

\subsection{R wrapper}
The {\tt TESS3} program can be executed using a wrapper in R software environment. The wrapper and helper functions are defined in the R script \verb|src/Rwrapper/TESS3.R|, the user can directly source this script in a R session. We now present function define in this script: 


\begin{itemize}

\item \verb|TESS3|

\paragraph{Description}
The wrapper function that call the command-line program. This function create a directory \verb|TESS3_workingDirectory| to store input and output file.
\paragraph{Usage}
\begin{Verbatim}
project = TESS3( genotype,
                 spatialData,
                 K,
                 ploidy=1,
                 seed=-1, 
                 rep = 1, 
                 maskedProportion = 0.0, 
                 alpha = 0.001 )
\end{Verbatim}
\paragraph{Arguments}
\begin{itemize}
\item \verb|genotype|: genotype R matrix of size $n$ individual by $L$ loci or the .geno format file name.
\item \verb|spatialData|: coordinate R matrix of size $n$ individual by $2$ or the .coord format file name.
\item \verb|K|: vector of number of ancestral cluster. The {\tt TESS3} program is run for each element of this vector.
\item \verb|ploidy|: 1 if haploid, 2 if diploid.
\item \verb|rep|: number of run for each number of ancestral population.
\item \verb|maskedProportion|: if \verb|maskedProportion > 0|, the cross-entropy criterion is calculated for this percentage of masked genotypes.
\item \verb|alpha|: value of the normalized regularization parameter. This parameter control the spatial regularity of the ancestry estimates.
\end{itemize}

\item \verb|Getter|

\paragraph{Description}
Functions useful to fetch results of \verb|TESS3| R function.
\paragraph{Usage}
\begin{Verbatim}
getQ( project, K, run = "best" )
 
getG( project, K, run = "best" )
  
getFst( project, K, run = "best" )
  
getCrossEntropy( project, func = mean ) 

getLeastSquared( project, func = mean )
\end{Verbatim}
\paragraph{Arguments}
\begin{itemize}
\item \verb|project|: object returned by the \verb|TESS3| R function.
\item \verb|K|: number of ancestral cluster.
\item \verb|run|: number of the run, or \verb|"best"| if you want the best result with respect to the least-squared criterion.
\item \verb|func|: function used to summarize data over all run.

\end{itemize}

\item \verb|Reader|

\paragraph{Description}
Function useful to read data from file.
\paragraph{Usage}
\begin{verbatim}
read.coord( file )
\end{verbatim}
\paragraph{Arguments}
\begin{itemize}
\item \verb|file| name of the file to read.
\end{itemize}


\end{itemize}

\noindent
A full example in R is available at the end of this note.



\section{Tutorial}

\subsection{Data set}
The data set that we analyse in this tutorial is a simulated dataset from a {\it Arabidopsis thaliana} Data set used in~\cite{atwell2010genome}. We obtain a haploid data set of $n = 170$ individual and $L = 26943$ loci.

\subsection{Example of analysis using {\tt TESS3}}

This tutorial example describe how to run {\tt TESS3} in R for different values of $K$ the number of ancestral populations. Then we plot the cross entropy criterion with respect to $K$. Finally we use R script available on {\tt POPS} website (see http://membres-timc.imag.fr/Olivier.Francois/pops.html) to plot the ancestry coefficient with the $K$ found with the cross entropy criteria~\cite{jay2012forecasting}. You need to change \verb|TESS3_directory| and \verb|POPS_directory| by the directory path of {\tt TESS3} and {\tt POPS}.

\begin{Verbatim}[frame=single]
source("TESS3_directory/src/Rwrapper/TESS3.R")
library(LEA)

setwd( "TESS3_directory/data/simulated/Athaliana" )
###########################################################################
# Run TESS3 on a data set simualted from an Arabidopsis Athalina data set #
###########################################################################

#read data
spatialData = read.coord("Athaliana.coord")
n = nrow(spatialData)

project = TESS3( genotype = "Athaliana.geno", 
                 spatialData = "Athaliana.coord", 
                 K = 1:5, 
                 ploidy = 1, 
                 rep = 1, 
                 maskedProportion = 0.2)


###########################################
# Chose of K with cross-entropy criterion #
###########################################

plot( 1:5, 
      getCrossEntropy( project ), 
      main  = "Cross entropy",
      type="b", 
      xlab = "K", 
      ylab = "corss entropy" )

################################
# Plot result on map for K = 3 #
################################

# R script available on http://membres-timc.imag.fr/Olivier.Francois/pops.html
source("POPS_direction/R/POPSutilities.r")

asciiFile="/home/cayek/Projects/TESS3/data/simulated/Athaliana/down_etopo1.asc"
grid=createGridFromAsciiRaster(asciiFile)
# To display only altitudes above 0:
constraints=getConstraintsFromAsciiRaster(asciiFile,cell_value_min=0)

maps(matrix = getQ( project, K = 3 ),
     coord = spatialData,
     grid=grid,constraints=constraints,method="max",main="ancestry coefficient with K = 3")

\end{Verbatim}



\subsection{Example of genome-scan for selection using {\tt TESS3}}
This tutorial show an example in R of use of the ancestral allele frequency differentiation statistic computed by {\tt TESS3}. We first run {\tt TESS3} with $K = 3$ ancestral populations. Then using standard population genetic theory the $F_{\tt ST}$-statistic is transformed into squared t-score and p-value are computed using a fisher distribution. The test inflation due to neutral population structure is corrected by recalibrating the $t^2$-score to have uniform distribution under the null hypothesis. Finally, we control the false discovery rate using a Benjamini Hochberg procedure and plot the Manhattan plot. You need to change \verb|TESS3_directory| by the directory path of {\tt TESS3}.

\begin{Verbatim}[frame=single]
source("TESS3_directory/src/Rwrapper/TESS3.R")
library(LEA)

setwd( "TESS3_directory/data/simulated/Athaliana" )
###########################################################################
# Run TESS3 on a data set simualted from an Arabidopsis Athalina data set #
###########################################################################

#read data
spatialData = read.coord("Athaliana.coord")
genotype = read.geno("Athaliana.geno")
n = nrow(spatialData)

project = TESS3( genotype = genotype, 
                 spatialData = spatialData, 
                 K = 3, 
                 ploidy = 1, 
                 rep = 5 )


#############################
# Genome scan for selection #
#############################

#### Fst with TESS3 
Fst = getFst( project, K = 3 )
Fst[Fst < 0.0] = 0.0

#### Convert Fst into t score
squared.t.scores = Fst*(n-2)/(1-Fst)

#### recalibrated p-values
gif = 25
adj.p.values = pf( squared.t.scores/gif , df1 = 2, df2 = n-3, lower = FALSE )

hist(adj.p.values,prob=TRUE)

#### Benjamini Hochberg procedure
alpha = 1e-10
L = length(adj.p.values)
# return a list of candidates with an expected FDR of alpha.
w = which(sort(adj.p.values) < alpha * (1:L) / L)
candidates = order(adj.p.values)[w]
limite = max(adj.p.values[candidates])

#### Manhattan plot 
plot( 1:length(adj.p.values),-log10(adj.p.values) , 
      main = "Manhattan Plot" , 
      xlab = "indices", 
      ylab="- Log P-value", 
      pch=19, cex = .5) 
#add limite
abline( -log10(limite), 0, col = "green", lty = 6, lwd = 3 )
\end{Verbatim}

\section{Contact}
If you need assistance, do not hesitate to send us an email (kevin.caye@imag.fr or olivier.francois@imag.fr). 

\bibliographystyle{plain}
\bibliography{biblio}

\end{document}
