\documentclass[10pt,a4paper]{article}

\usepackage{amssymb}
\usepackage[french,english]{babel}
\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{lineno}
\usepackage{cite}
\usepackage{float}
\usepackage{ccaption}
\usepackage{caption}
\usepackage{array}
\usepackage{lscape}
\usepackage[hmargin=2cm,vmargin=2cm]{geometry}
\usepackage{fancyvrb}
\usepackage{hyperref}

\title{\bf \Large A short manual for {\tt TESS3}:
a program to estimate spatial population structure\\
\large (command-line and R wrapper)
}

\author{
        Kevin Caye (kevin.caye@imag.fr)\\
        Olivier FranÃ§ois (olivier.francois@imag.fr)\\
}

\newcommand{\bp}{\mathbf{p}}
\newcommand{\LLL}{\mathcal{L}}

%% BEGIN DOC
\begin{document}


\maketitle
\begin{center}
{\it Please, print this reference manual only if it is necessary.}
\end{center}

\noindent
This short manual aims to help users to run {\tt TESS3} command-line engine on Mac and Linux. 

\section{Description} 

Inference of spatial population structure are commonly perform with computer program based on intensive stochastic simulation. These methods do not scale with the dimension of the data sets generated from next-generation sequencing technologies. The computer program {\tt TESS3} has functionalities similar to the spatial bayesian clustering program  {\tt TESS}~\cite{durand2009spatial}, but has run-times several order faster than those of {\tt TESS} 2.3. The method is based on geographically constrained non-negative matrix factorization, and provides similar results that with {\tt TESS} 2.3. In addition {\tt TESS3} computes an ancestral allele frequency differentiation statistic that can be used to perform selection scans.

\section{Installation} 

\noindent
TO DO 

\section{Data format}

\subsection{input file}
The {\tt sNMF} input file consists of a genotype file in the {\bf geno} format and coordinate file in the {\bf coord} format. 
\begin{itemize}
\item {\bf geno} (example.geno)

The {\bf geno} format has one row for each SNP.
  Each row contains 1 character per individual:
  0 means zero copies of the reference allele.
  1 means one copy of the reference allele.
  2 means two copies of the reference allele.
  9 means missing data.

Below, an example of a geno file for $n=3$ individuals and $L=4$ loci.
\begin{center}
\footnotesize
\begin{Verbatim}[frame=single]
112
010
091
121
\end{Verbatim}
\end{center}


\item {\bf coord} (example.coord)

The {\bf coord} format has one row for each individual. Each row contain the \verb|longitude latitude| information of the individual.

Below, an example of a geno file for $n=3$ individuals and $L=4$ loci.
\begin{center}
\footnotesize
\begin{Verbatim}[frame=single]
2.515455200203690111e+01 5.439077948590419709e+01
-8.429345306090160861e+00 4.019713388759510053e+01
1.351129055178800087e+01 5.585331731335860184e+01
\end{Verbatim}
\end{center}

\end{itemize}

\noindent
Other formats for the genotype data sets can be used thanks to the {\tt LEA} R package~\cite{frichot2015lea}. Indeed, {\tt LEA} enable to convert into {\bf geno} format the following usual formats : {\bf ped}, {\bf ancestrymap}, {\bf vcf} and {\bf lfmm}.

\subsection{output files}
\noindent
There are three main {\bf output files}.

\begin{itemize}
\item The file with the extension the {\bf .Q} contains individual admixture coefficients.
It contains a matrix with $n$ rows (the number of individuals) and $K$ columns (the 
number of ancestral populations).
\item The file with the extension {\bf .G} contains the ancestral genotypic frequencies.
It contains a matrix with $n_a\times L$ lines (the number of alleles times the number of SNPs) 
and $K$ columns (the number of ancestral populations). 
For a diploid SNP, the first line contains the ancestral frequencies for the number 
of allele equals to 0, the second line contains the ancestral frequencies for the 
number allele equals to 1, the third line contains the ancestral frequencies for 
the number of alleles equal to 2.
\item The file with the extension {\bf .Fst} contains ancestral allele frequency differentiation statistic. In this file each row is the statistic estimate for each loci. 
\end{itemize}

There also are less important {\bf output files}, but that can be useful to have information on the graph computed, the least-squared criterion and the cross-entropy criterion.

\begin{itemize}
\item The file with the extension the {\bf .W} contains the edge weight matrix of the nearest-neighbor graph computed from coordinates data.
\item The file with the extension the {\bf .sum} contains the value of least-squared criterion, the cross entropy criteria on all data and only on masked data if the user asked it.
\end{itemize}

\section{Run the program}
\subsection{Command-line}
The {\tt TESS3} program can be executed from a command line. The format is:
\begin{Verbatim}[frame=single]
./TESS3 -x genotype_file.geno -K number_of_ancestral_populations  -r coordinates_file
\end{Verbatim}

\noindent
Only these three options are mandatory. There is no ordering for the options in the command line. 
Here is a description of these options:
\begin{itemize}
\item \verb|-x genotype_file.geno| is the path to the genotype file (in .geno format).
\item \verb|-K number_of_ancestral_populations| is the number of ancestral populations. 
\item \verb|-r coordinates_file| is the path to the coordinate file (in .coord format). 
\end{itemize}

\noindent
Additional options are available:
\begin{itemize}
\item \verb|-a alpha| is the value of the normalized regularization parameter (by default: 0.001). This parameter control the spatial regularity of the ancestry estimates.
\item \verb|-W edge_weight_input| is the path to the file that contains a edge weight matrix of the spatial graph. The file has to contain each element of the matrix separates by a space row by row. The program use this graph in place of the one computed if no edge weight input file is given.
\item \verb|-q output_Q| is the path for the output file containing the ancestry coefficients. By default, the name of the output file is the same name as the input file with the extension .K.Q.
\item \verb|-g output_G| is the path to the output file containing the ancestral genotype frequencies. By default, the name of the output file is the same name as the input genotype file with the extension .K.G.
\item \verb|-f output_FST| is the path to the output file containing the ancestral allele frequency differentiation statistic. By default, the name of the output file is the same name as the input genotype file with the extension .K.Fst.
\item \verb|-y output_FST| is the path to the output file containing the value of least-squared criterion, the cross entropy criteria on all data and only on masked data. By default, the name of the output file is the same name as the input genotype file with the extension .K.sum.
\item \verb|-c perc| is the percentage of masked genotypes. If this option is set, the cross-entropy criterion is calculated (see\cite{frichot2014fast} for more details on the cross-entropy criterion). The default percentage is $5\%$.
\item \verb|-e tolerance| is the tolerance error in the {\tt sNMF} optimization algorithm (by default: 0.0000001). 
\item \verb|-i iteration_number| is the max number of iterations of the algorithm (default: 200). 
\item \verb|-I nb_SNPs| starts the algorithm with a run of sNMF using a subset of nb\_SNPs random SNPs. This option can speed up sNMF estimation for very large data sets.
\item \verb|-Q input_Q| is the path to an initial file for the $Q$ matrix containing individual admixture coefficients. If both \verb|-I| and \verb|-Q| are set, \verb|-Q| is chosen.
\item \verb|-s seed| is a seed to initialize the random number generator. 
\item \verb|-m ploidy|  1 if haploid, 2 if diploid (default: 2). 
\item \verb|-p p| is the number of CPUs to use when the algorithm is run on a multiprocessor system.
Be aware that the number of processes has to be lower or equal to the number 
of CPU units available on your computer (default: 1).

\end{itemize}


\noindent
If you need a summary of options, you can use the \verb|-h| option by typing the following command
\footnotesize
\begin{Verbatim}[frame=single]
./bin/sNMF -h
\end{Verbatim}
\noindent
\normalsize

\subsection{R wrapper}
The {\tt TESS3} program can be executed using a wrapper in R software environment. The wrapper and helper functions are defined in the R script \verb|src/Rwrapper/TESS3.R|, the user can directly source this script in a R session. We now present function define in this script: 


\begin{itemize}

\item \verb|TESS3|

\paragraph{Description}
The wrapper function that call the command-line program. This function create a directory \verb|TESS3_workingDirectory| to store input and output file.
\paragraph{Usage}
\begin{Verbatim}
project = TESS3( genotype,
                 spatialData,
                 K,
                 ploidy=1,
                 seed=-1, 
                 rep = 1, 
                 maskedProportion = 0.0, 
                 alpha = 0.001 )
\end{Verbatim}
\paragraph{Arguments}
\begin{itemize}
\item \verb|genotype|: genotype R matrix of size $n$ individual by $L$ loci or the .geno format file name.
\item \verb|spatialData|: coordinate R matrix of size $n$ individual by $2$ or the .coord format file name.
\item \verb|K|: vector of number of ancestral cluster. The {\tt TESS3} program is run for each element od the vector.
\item \verb|ploidy|: 1 if haploid, 2 if diploid
\item \verb|rep|: number of run for each number of ancestral population.
\item \verb|maskedProportion|: if \verb|maskedProportion > 0|, the cross-entropy criterion is calculated for this percentage of masked genotypes
\item \verb|alpha|: value of the normalized regularization parameter. This parameter control the spatial regularity of the ancestry estimates.
\end{itemize}

\item \verb|Getter|

\paragraph{Description}
Functions useful to fetch results of \verb|TESS3| R function.
\paragraph{Usage}
\begin{Verbatim}
getQ( project, K, run = "best" )
 
getG( project, K, run = "best" )
  
getFst( project, K, run = "best" )
  
getCrossEntropy( project, func = mean ) 

getLeastSquared( project, func = mean )
\end{Verbatim}
\paragraph{Arguments}
\begin{itemize}
\item \verb|project|: object returned by the \verb|TESS3| R function.
\item \verb|K|: number of ancestral cluster.
\item \verb|run|: number of the run, or \verb|"best"| if you want the best result with respect to the least-squared criterion.
\item \verb|func|: function used to summarize data over all run.

\end{itemize}

\item \verb|Reader|

\paragraph{Description}
Function useful to read data from file.
\paragraph{Usage}
\begin{verbatim}
read.coord( file )
\end{verbatim}
\paragraph{Arguments}
\begin{itemize}
\item \verb|file| name of the file to read.
\end{itemize}


\end{itemize}

\noindent
A full example in R is available at the end of this note.



\section{Tutorial}

\subsection{Data set}
The data set that we analyze in this tutorial is an Asian human data set.
This data is a worldwide sample of genomic DNA (10757 SNPs) from 934 individuals,
taken from the Harvard Human Genome Diversity Project - Centre
Etude Polymorphism Humain (Harvard HGDP-CEPH). 
In those data, each marker has been ascertained in samples of Mongolian
ancestry (referenced population HGDP01224) \cite{Patterson_2012}. 

\subsection{Example}
\paragraph{Create a data set with masked data}

In the main directory, type:
\begin{Verbatim}[frame=single]
./bin/createDataSet -x examples/panel11.geno
\end{Verbatim}
\noindent
output for createDataSet
\begin{Verbatim}[frame=single]
summary of the options:

        -n (number of individuals)                 934
        -L (number of loci)                        10757
        -s (seed random init)                      11162993670188721480
        -r (percentage of masked data)             0.05
        -x (genotype file)                         examples/panel11.geno
        -o (output file)                           examples/panel11_I.geno
        - diploid

Write genotype file with masked data examples/panel11_I.geno:		OK.
\end{Verbatim}
\noindent
A file with 5 \% of masked data \verb|examples/panel11_I.geno| has been created.

\paragraph{Run {\tt sNMF}}

Then, run {\tt sNMF} with 5 \% of masked data and $K=5$
\begin{Verbatim}[frame=single]
./bin/sNMF -x examples/panel11_I.geno -K 5
\end{Verbatim}
\noindent
The output for sNMF is
\begin{Verbatim}[frame=single]
./bin/sNMF -x examples/panel11_I.geno -K 5 
summary of the options:

        -n (number of individuals)             934
        -L (number of loci)                    10757
        -K (number of ancestral pops)          5
        -x (input file)                        examples/panel11_I.geno
        -q (individual admixture file)         examples/panel11_I.5.Q
        -g (ancestral frequencies file)        examples/panel11_I.5.G
        -i (number max of iterations)          200
        -a (regularization parameter)          0
        -s (seed random init)                  11162857829069388553
        -e (tolerance error)                   0.0001
        -p (number of processes)               1
        - diploid

Read genotype file examples/panel11_I.geno:		OK.

Main algorithm:
[                                                                           ]
[===============================================]
Number of iterations: 142

Least-square error: 4597146.245722
Write individual ancestry coefficient file examples/panel11_I.5.Q:		OK.
Write ancestral allele frequency coefficient file examples/panel11_I.5.G:	OK.
\end{Verbatim}

\noindent
The results files \verb|examples/panel11_I.Q| and \verb|examples/panel11_I.G| have been created.
{\tt sNMF} also dislpays the number of iterations and the least-squares error: $||X - QG||_F^2$
(see [1]).

\paragraph{Compute the cross-entropy criterion}

Finally, calculate the cross-entropy criterion:
\begin{Verbatim}[frame=single]
./bin/crossEntropy -x examples/panel11.geno -K 5
\end{Verbatim}
\noindent
log for crossEntropy:
\begin{Verbatim}[frame=single]
summary of the options:

        -n (number of individuals)         934
        -L (number of loci)                10757
        -K (number of ancestral pops)      5
        -x (genotype file)                 examples/panel11.geno
        -q (individual admixture)          examples/panel11_I.5.Q
        -g (ancestral frequencies)         examples/panel11_I.5.G
        -i (with masked genotypes)         examples/panel11_I.geno
        - diploid

Cross-Entropy (all data):	 0.746568
Cross-Entropy (masked data):	 0.756867
\end{Verbatim}
\noindent
The {\tt crossEntropy} program displays the cross-entropy calculated for all data and for the masked data.
The cross-entropy for all data is always lower than the cross-entropy for the masked data. 
The value useful to compare runs is the {\bf cross-entropy for the masked data}.


\paragraph{All in once}
We can also run all in once using the \verb|-c| option:
\begin{Verbatim}[frame=single]
./bin/sNMF -x examples/panel11.geno -K 5 -c
\end{Verbatim}

\subsection{How to choose $K$}

To choose a value for $K$, we launch {\tt sNMF} for each value of $K$ from 2 to 10. The Figure displays the cross-entropy value obtained for each value of $K$. $K=6$ is the best value according to the cross-entropy criterion because the cross-entropy does not decrease for $K$ greater than 6. This analysis can be completed by displaying the $Q$-matrices associated with each run. 
\begin{Verbatim}[frame=single]
for K in 1 2 3 4 5 6 7 8 9 10 
do 
	echo K=K; ./bin/sNMF -x examples/panel11.geno -K K -c > examples/panel11.K.log
done
grep "Cross-Entropy (masked data):" examples/panel11.*.log
\end{Verbatim}

%\centerline{\includegraphics[width=15cm]{ce.pdf}}
\noindent{\bf Figure.}  {\bf Values of the cross-entropy criterion for 10 {\tt sNMF} runs (dataset HGDP01224).} 
\\
\\

\noindent
The ancestry coefficients for $K=6$ are available in the file examples/panel11.6.Q. The format of the file containing the ancestry coefficients is the format used by the software {\tt STRUCTURE}. This file can be used as input into CLUMPP, the software to gather output runs [3] and into distruct to display admixture plots [4].\\
\\
\noindent
Below, the first three lines of examples/panel11.6.Q.
\begin{center}
\footnotesize
\begin{Verbatim}[frame=single]
0.150569 0.0135955 0.0103463 0.779787 0.00755981 0.0381422
0.156673 0.0584628 0.0543375 0.682024 9.9991E-05 0.0484029
0.280393 0.00873713 0.0120376 0.635205 0.0635271 9.9991E-05
\end{Verbatim}
\end{center}
{\it Tips:} It is clearly useful to perform several runs of {\tt sNMF} for the same value of $K$ and to choose the run with the smallest value of the cross-entropy criterion.



\section{Contact}
If you need assistance, do not hesitate to send us an email (eric.frichot@imag.fr or olivier.francois@imag.fr). 
A FAQ (Frequently Asked Questions) section is available 
on our webpage (http://membres-timc.imag.fr/Olivier.Francois/snmf.html). 
{\tt sNMF} software is still under development. All your comments and feedbacks are more than welcome.\\
\\
\noindent
[1] Eric Frichot, FranÃ§ois Mathieu, ThÃ©o Trouillon, Guillaume Bouchard, Olivier FranÃ§ois. (2014) {\it Fast and efficient estimation of individual ancestry coefficients}. Genetics 196: 973 -- 983.\\ 
\\
\noindent
[2]  Nick J. Patterson, Priya Moorjani, Yontao Luo, Swapan Mallick, Nadin Rohland, Yiping Zhan, Teri
Genschoreck, Teresa Webster, and David Reich. (2012) {\it Ancient admixture in human history}. Genetics
192: 1065 -- 1093.\\
\\
\noindent
[3] Mattias Jakobsson, Noah A. Rosenberg. (2007) {\it CLUMPP: a cluster matching and permutation program for dealing with label switching and multimodality in analysis of population structure}. Bioinformatics, 23(14): 1801--1806.
\\
\\
\noindent
[4] Noah A. Rosenberg. (2004) {\it DISTRUCT: a program for the graphical display of population structure}. Molecular Ecology Notes, 4(1): 137--138.


\bibliographystyle{plain}
\bibliography{biblio}

\end{document}
